File: lingting-core/src/test/java/live/lingting/framework/retry/RetryTest.java
package live.lingting.framework.retry;

import live.lingting.framework.function.ThrowingSupplier;
import lombok.SneakyThrows;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import java.time.Duration;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * @author lingting 2023-10-24 14:33
 */
class RetryTest {

	@SneakyThrows
	@Test
	void test() {
		int expected = 3;

		AtomicInteger atomic = new AtomicInteger(0);
		ThrowingSupplier<Integer> supplier = () -> {
			int i = atomic.get();
			if (i == expected) {
				return i;
			}
			atomic.set(i + 1);
			throw new IllegalStateException("异常");
		};

		Retry<Integer> retry = Retry.simple(4, Duration.ZERO, supplier);
		RetryValue<Integer> value = retry.value();
		Assertions.assertTrue(value.success());
		Assertions.assertEquals(expected, value.get());
		Assertions.assertEquals(4, value.logs().size());

		atomic.set(0);
		retry = Retry.simple(2, Duration.ZERO, supplier);
		value = retry.value();
		Assertions.assertFalse(value.success());
		Assertions.assertNull(value.value());
		Assertions.assertEquals(3, value.logs().size());

	}

}

File: lingting-core/src/test/kotlin/live/lingting/framework/retry/RetryTest.kt
package live.lingting.framework.retry

import java.time.Duration
import java.util.concurrent.atomic.AtomicInteger
import live.lingting.framework.function.ThrowingSupplier
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

/**
 * @author lingting 2023-10-24 14:33
 */
internal class RetryTest {
    @Test
    fun test() {
        val expected = 3

        val atomic = AtomicInteger(0)
        val supplier: ThrowingSupplier<Int> = ThrowingSupplier<Int> {
            val i = atomic.get()
            if (i == expected) {
                return@ThrowingSupplier i
            }
            atomic.set(i + 1)
            throw IllegalStateException("异常")
        }

        var retry: Retry<Int> = Retry.simple(4, Duration.ZERO, supplier)
        var log = retry.last()
        assertTrue(log.ex == null)
        assertEquals(expected, log.value)
        assertEquals(4, retry.logs.size)

        atomic.set(0)
        retry = Retry.simple(2, Duration.ZERO, supplier)
        log = retry.last()
        assertFalse(log.ex == null)
        assertNull(log.value)
        assertEquals(3, retry.logs.size)
    }
}
